### PR TITLE
Better syntactic sugar

### PR DESCRIPTION
**Pull request by dannyech**

### CODE REVIEW

File: compiler/analyzer.ml [hunk 1]
```diff
@@ -56,19 +56,37 @@ let builtins = VarMap.add "mult_dist" {
 
 let builtins = VarMap.add "shift_dist" {
   name = "shift_dist";
-  s_type = Func({ param_types = [Dist_t; Num]; return_type = Dist_t; });
+  s_type = Func({ param_types = [Num; Dist_t]; return_type = Dist_t; });
```
Comment (alexandramedway) [inline compiler/analyzer.ml]: TODO, make sure this correct

File: compiler/analyzer.ml [hunk 2]
```diff
@@ -56,19 +56,37 @@ let builtins = VarMap.add "mult_dist" {
...
 let builtins = VarMap.add "exp_dist" {
   name = "exp_dist";
-  s_type = Func({ param_types = [Dist_t; Num]; return_type = Dist_t; });
+  s_type = Func({ param_types = [Num; Dist_t]; return_type = Dist_t; });
+  builtin = true;
+} builtins
+
+let builtins = VarMap.add "sample_dist" {
+  name = "sample_dist";
+  s_type = Func({ param_types = [Num; Dist_t]; return_type = List(Num); });
```
Comment (alexandramedway) [inline compiler/analyzer.ml]: This is num dist, the ones below are dist num. @dannyech are you fixing this?

File: compiler/analyzer.ml [hunk 3]
```diff
@@ -135,8 +160,8 @@ let str_of_unop = function
 let str_of_binop = function
   (* Dist *)
   | D_Plus -> "<+>" | D_Times -> "<*>"
-  | D_Shift -> ">>" | D_Stretch -> "<>"
-  | D_Power -> "^^"
+  | D_Shift -> "|+" | D_Stretch -> "|*"
+  | D_Sample -> "<>" | D_Power -> "|**"
```
Comment (alexandramedway) [inline compiler/analyzer.ml]: nit: spacing

File: compiler/analyzer.ml [hunk 4]
```diff
@@ -524,80 +550,57 @@ and check_unop env op e =
...
+        | D_Stretch -> "stretch_dist"
+        | D_Power -> "exp_dist"
+        | D_Sample -> "sample_dist"
+        | _ -> dead_code_path_error "check_binop" in
+      
+    (* Unsugar expression and refeed it to Analyzer *)
+    let unsugared = Ast.Call(Ast.Id(func_name), [e1; e2]) in
+    check_expr env unsugared
+  
+  else (* binop is not sugar. Check, constrain, and proceed. *)
```
Comment (alexandramedway) [inline compiler/analyzer.ml]: This is the same as before?

File: compiler/lib/core.py [hunk 5]
```diff
@@ -53,36 +54,42 @@ def make_dist(start, end, f):
...
-    start1 = random.randint(0, 99)
-    s1 = d1[start1::SAMPLE_STEP]
-    start2 = random.randint(0, 99)
-    s2 = d2[start2::SAMPLE_STEP]
-    add = []
-    for i in s1:
-        for j in s2:
-            add.append(i + j)
-    return sorted(add)
+    """Return the sum of two distributions, adding each combination"""
```
Comment (alexandramedway) [inline compiler/lib/core.py]: this is good

File: compiler/lib/core.py [hunk 6]
```diff
@@ -53,36 +54,42 @@ def make_dist(start, end, f):
...
+def exp_dist(n, d):
+    """Exponentiate distribution d, raising each element to power n"""
     return [ x ** n for x in d ]
 
+def sample_dist(n, d):
+    """Return a random sample of n elements in distribution d"""
+    return sorted([ random.randint(0, DIST_LENGTH - 1) for x in range(n) ])
+
+def prob(d, n):
+    return len([i for i in d if i < n]) / DIST_LENGTH
```
Comment (alexandramedway) [inline compiler/lib/core.py]: add comments for these too for consistency. see above.

File: compiler/lib/core.py [hunk 7]
```diff
@@ -53,36 +54,42 @@ def make_dist(start, end, f):
...
+
+def expected(d):
```
Comment (alexandramedway) [inline compiler/lib/core.py]: Comment here also

File: compiler/lib/core.py [hunk 8]
```diff
@@ -53,36 +54,42 @@ def make_dist(start, end, f):
...
+    return sum(d) / DIST_LENGTH
+
+def concat_str(s1, s2):
```
Comment (alexandramedway) [inline compiler/lib/core.py]: Comment here

File: compiler/scanner.mll [hunk 9]
```diff
@@ -37,7 +37,11 @@ rule token = parse
 
 (* Dist Operators *)
 | "<+>" { DPLUS }   | "<*>" { DTIMES }
-| "^^"  { DPOWER }
```
Comment (alexandramedway) [inline compiler/scanner.mll]: good changes

File: compiler/analyzer.ml [hunk 10]
```diff
@@ -56,19 +56,37 @@ let builtins = VarMap.add "mult_dist" {
...
+  builtin = true;
+} builtins
+
+let builtins = VarMap.add "sample_dist" {
+  name = "sample_dist";
+  s_type = Func({ param_types = [Num; Dist_t]; return_type = List(Num); });
+  builtin = true;
+} builtins
+
+let builtins = VarMap.add "prob" {
```
Comment (akalicki) [inline compiler/analyzer.ml]: aren't we calling this `P`?

File: compiler/analyzer.ml [hunk 11]
```diff
@@ -56,19 +56,37 @@ let builtins = VarMap.add "mult_dist" {
...
+  name = "prob";
+  s_type = Func({ param_types = [Dist_t; Num]; return_type = Num; });
+  builtin = true;
+} builtins
+
+let builtins = VarMap.add "expected" {
```
Comment (akalicki) [inline compiler/analyzer.ml]: aren't we going to call this `E`?

File: test/lib/_dist.ods [hunk 12]
```diff
@@ -1,4 +1,4 @@
 do square = (x) -> return x * x
 do x = <0, 1> | square |
 
-do y = <5 + 6, 7 + 8> | (x) -> return x * x |
+/*do y = <5 + 6, 7 + 8> | (x) -> return x * x |*/
```
Comment (akalicki) [inline test/lib/_dist.ods]: uncomment

File: compiler/scanner.mll [hunk 13]
```diff
@@ -38,7 +38,11 @@ rule token = parse
...
+| "|**" { DPOWER }  | "|*"  { DSTRETCH }
+| "|+"  { DSHIFT }
+
+(* List Operator *)
+| "::"    { CONS }
```
Comment (akalicki) [inline compiler/scanner.mll]: can you add a compiler test for this?

File: test/scanner/tokenize.ml [hunk 14]
```diff
@@ -11,11 +11,13 @@ let stringify = function
...
   | COMMA -> "COMMA"    | VBAR -> "VBAR"
-  | DDELIM -> "DDELIM"
+  | DDELIM -> "DDELIM"  | DISC -> "DISC"
 
   (* Dist Operators *)
   | DPLUS -> "DPLUS"    | DTIMES -> "DTIMES"
-  | DPOWER -> "DPOWER"
+  | DPOWER -> "DPOWER"  | DSHIFT -> "DSHIFT"
+  | DSTRETCH -> "DSTRETCH"
+
```
Comment (akalicki) [inline test/scanner/tokenize.ml]: nit: remove extra space

**Pull request merged**